Doc and example code snippet for `mlflow.pyfunc.spark_udf` restoring model dependencies (#5677)
mlflow/mlflow: docs/source/models.rst

For models where no schema is defined, no changes to the model inputs and outputs are made. MLflow will
propogate any errors raised by the model if the model does not accept the provided input type.

> The model might be trained in a different python environment which mismatch with current python environment,
> In this case, when calling :py:func:`mlflow.pyfunc.load_model`, a warning message will be printed
> saying which python packages are mismatched. In order to get full dependencies of the model, you can
> call :py:func:`mlflow.pyfunc.get_model_dependencies`. Furthermore, if you want to run model inference
> in the same environment used in model training, you can call
> :py:func:`mlflow.pyfunc.spark_udf` with `env_manager` argument set as "conda" which return a python UDF,
> and the UDF runs inference in the training model python environment restored by Conda.

Reviewer: ```suggestion
The python environment that a PyFunc model is loaded into for prediction or inference may differ from the environment in which it was trained. In the case of an environment mismatch, a warning message will be printed when calling :py:func:`mlflow.pyfunc.load_model`. This warning statement will identify the packages that have a version mismatch between those used during training and the current environment.  In order to get the full dependencies of the environment in which the model was trained, you can call :py:func:`mlflow.pyfunc.get_model_dependencies`. Furthermore, if you want to run model inference
in the same environment used in model training, you can call
:py:func:`mlflow.pyfunc.spark_udf` with the `env_manager` argument set as "conda". This will generate the environment from the `conda.yaml` file, ensuring that the python UDF will execute with the exact package versions that were used during training. 
```

    df = spark_df.withColumn("prediction", pyfunc_udf(struct("name", "age")))

If you want to use conda to restore the python environment that was used to train the model,
set the `env_manager` argument when calling :py:func:`mlflow.pyfunc.spark_udf`.

.. rubric:: Example

.. code-block:: py

    from pyspark.sql.types import ArrayType, FloatType
    from pyspark.sql.functions import struct

    pyfunc_udf = mlflow.pyfunc.spark_udf(

>         "path/to/model", result_type=ArrayType(FloatType()),

Reviewer 2: ```suggestion
        spark_session, "path/to/model", result_type=ArrayType(FloatType()),
```

Reviewer 2: Can we also fix line 1912?

        env_manager="conda"  # Use conda to restore the environment used in training

>     )

Reviewer 2: Is this line correct? The first argument of `spark_udf` is

Reviewer 3: We're missing the `spark` argument, which is the first argument to the function. See https://github.com/mlflow/mlflow/pull/5677/files#diff-27092883ef9a4a81ab9b5ea59442ab48dd314babd61428270948fcfae0a80f28R533-R541.

Reviewer 2: The example above this one also has the same issue. Let's fix it.

Reviewer 2: Can we add `import mlflow` so we can copy the code in a script and run it?

>     pyfunc_udf = mlflow.pyfunc.spark_udf(
>         "path/to/model", result_type=ArrayType(FloatType()),
>         env_manager="conda"  # Use conda to restore the environment used in training
>     )

Reviewer 3: ```suggestion
    pyfunc_udf = mlflow.pyfunc.spark_udf(
        spark,
        "path/to/model", 
        result_type=ArrayType(FloatType()),
        env_manager="conda"  # Use conda to restore the environment used in training
    )
```

Reviewer 2: Can we include the following

```python
from pyspark.sql import SparkSession

spark = SparkSession.builder.getOrCreate()
```

in the example?

Reviewer 2: Can we add `import mlflow` so we can copy the code in a script and run it?

Reviewer 2: ```python
    from pyspark.sql.types import ArrayType, FloatType
    from pyspark.sql.functions import struct
    from pyspark.sql import SparkSession
    import mlflow

    spark = SparkSession.builder.getOrCreate()
    pyfunc_udf = mlflow.pyfunc.spark_udf(
        spark,
        "path/to/model",
        result_type=ArrayType(FloatType()),
        env_manager="conda",  # Use conda to restore the environment used in training
    )
    df = spark_df.withColumn("prediction", pyfunc_udf(struct("name", "age")))
```

Reviewer 2: Ah, `spark_df` is still undefined.

Reviewer 3: Defining a model would make it even longer. I think this is good enough.

Reviewer 2: Sounds good :)
